anta/tools.py

```diff
@@ -415,3 +418,45 @@ def format_data(data: dict[str, bool]) -> str:
     "Advertised: True, Received: True, Enabled: True"
     """
     return ", ".join(f"{k.capitalize()}: {v}" for k, v in data.items())
+
+
+async def limit_concurrency(coroutines: AsyncIterator[Coroutine[Any, Any, T]], limit: int) -> AsyncIterator[asyncio.Task[T]]:
+    """Schedule a limited number of coroutines concurrently.
+
+    Inspired by: https://death.andgravity.com/limit-concurrency
+
+    Parameters
+    ----------
+    coroutines
+        An async iterator of coroutines.
+    limit
+        The maximum number of coroutines to run concurrently.
+
+    Yields
+    ------
+        Each completed task.
+    """
+    if limit <= 0:
+        msg = "Concurrency limit must be greater than 0."
+        raise RuntimeError(msg)
+
+    coros_ended = False
+    pending: set[asyncio.Task[T]] = set()
+
+    while pending or not coros_ended:
+        while len(pending) < limit and not coros_ended:
+            try:
+                # NOTE: The `anext` built-in function is not available in Python 3.9
+                coro = await coroutines.__anext__()  # pylint: disable=unnecessary-dunder-call
+            except StopAsyncIteration:  # noqa: PERF203
```

Comment:
lets add a comment that we don't want to stop the loop in case an exception is raised and thats why we ignore PERF203
```suggestion
            except StopAsyncIteration:  # noqa: PERF203
```

Reply:
Done.